### PR TITLE
Support OpenStack Swift Object Storage 

### PR DESCRIPTION
**Pull request by joec4i**
This adds support for OpenStack Swift. I've include unit test cases. 

This was also tested manually against a real Swift service.

```
INFO    [2022-04-05T15:29:51+10:00] Started /rs:fill/w:800/h:600/plain/swift:///tester/test/test.png  request_id=2Sr0AoRcFFo8TeLmSJ7W_ method=GET client_ip="::1"
INFO    [2022-04-05T15:29:51+10:00] Completed in 535.526079ms /rs:fill/w:800/h:600/plain/swift:///tester/test/test.png  request_id=2Sr0AoRcFFo8TeLmSJ7W_ method=GET status=200 processing_options="Width: 800; Height: 600; Format: png" client_ip="::1" image_url="swift:///tester/test/test.png"
```
Please review. Thanks 🙏

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -32,3 +33,5 @@ require (
 replace git.apache.org/thrift.git => github.com/apache/thrift v0.0.0-20180902110319-2566ecd5d999
 
 replace github.com/shirou/gopsutil => github.com/shirou/gopsutil v2.20.9+incompatible
+
+replace github.com/go-chi/chi/v4 => github.com/go-chi/chi v4.0.0+incompatible
```
Comment (joec4i) [inline go.mod:37]: Note: this is needed due to https://github.com/DataDog/dd-trace-go/issues/1220 . We can remove it once the new dd-trace-go release is out.

File: transport/swift/swift_test.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+	config.SwiftAuthURL = server.AuthURL
+	config.SwiftUsername = swifttest.TEST_ACCOUNT
+	config.SwiftAPIKey = swifttest.TEST_ACCOUNT
+	config.SwiftAuthVersion = 1
+
+	return server, err
+}
+
+func setupTestFile(t *testing.T) {
```
Comment (DarthSim) [inline transport/swift/swift_test.go:32]: Please rewrite tests to use `suite.Suite` from testify. This way you can initialize the server and upload the file once before all the tests.

**Review by DarthSim — CHANGES_REQUESTED** [review]
Hey Joe!
Thanks for your contribution! Please fix the commented things and I'll merge this

**Comment by DarthSim** [comment]
@Travis-Turner Could you check the docs, please?

File: transport/swift/swift.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+func parseObjectURL(url string) (container string, path string, err error) {
+	paths := strings.SplitN(strings.TrimPrefix(url, "/"), "/", 3)
+
+	if len(paths) != 3 {
+		return "", "", fmt.Errorf("invalid object url: %s. expecting {account}/{container}/{object_path}", url)
+	}
+
+	return paths[1], paths[2], nil
+}
```
[the remark this answers is not in the document]
Comment (joec4i) [reply transport/swift/swift.go:109]: That's a great point. Will update!

**Comment by joec4i** [comment]
Hi @DarthSim . Thanks for the quick review. I've updated the PR to address your feedback. Please have another look.

BTW I noticed that the CI job is not running all tests https://github.com/imgproxy/imgproxy/blob/49492c4/.github/workflows/test.yml#L33 is running `go test` instead of `go test ./...`  . Is that on purpose or an oversight? 

Thanks,
Joe

File: transport/swift/swift.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+	if err != nil {
+		return nil, fmt.Errorf("swift authentication error: %s", err)
+	}
+
+	return transport{con: c}, nil
+}
+
+func (t transport) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	container, path, err := parseObjectURL(req.URL.Path)
```
Comment (DarthSim) [inline transport/swift/swift.go:42]: Let's make it `swift://container/path`, not `swift:///container/path`.

```suggestion
	container := req.URL.Host
	path := req.URL.Path
```

You can trim the leading slash with `strings.trimPrefix(req.URL.Path, "/")` if this is required by Swift.

File: transport/swift/swift.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		return nil, fmt.Errorf("error opening object: %v", err)
+	}
+
+	header := make(http.Header)
+
+	if config.ETagEnabled {
+		if etag, ok := headers["Etag"]; ok {
+			header.Set("ETag", etag)
+
+			if len(etag) > 0 && etag == req.Header.Get("If-None-Match") {
```
Comment (DarthSim) [inline transport/swift/swift.go:70]: I believe we should close `object` here

File: transport/swift/swift_test.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+	assert.Nil(s.T(), err, "failed to set up transport")
+	assert.IsType(s.T(), transport, transport)
+}
+
+func (s *SwiftTestSuite) TestRoundTripWithETagDisabledReturns200() {
+	config.ETagEnabled = false
+	request, _ := http.NewRequest("GET", "swift:///test/test.png", nil)
+
+	transport, _ := New()
```
Comment (DarthSim) [inline transport/swift/swift_test.go:89]: It looks like we can initialize transport in `SetupSuite` as it doesn't change between tests.

**Review by DarthSim — CHANGES_REQUESTED** [review]

**Comment by DarthSim** [comment]
> BTW I noticed that the CI job is not running all tests https://github.com/imgproxy/imgproxy/blob/49492c4/.github/workflows/test.yml#L33 is running go test instead of go test ./... . Is that on purpose or an oversight?

My bad, fixed in the master. Thanks for catching this.

File: transport/swift/swift.go (hunk 4 above)
Comment (joec4i) [reply transport/swift/swift.go:42]: That was a bad commit from my test code. Apologies! Will fix that.

Comment (joec4i) [reply transport/swift/swift.go:42]: Sounds good.

A standard OpenStack Swift deployment has the resource path in the format of `/v1/{account}/{container}/{object}` according to https://docs.openstack.org/swift/latest/api/object_api_v1_overview.html . An example Swift object url may look like this `http://127.0.0.1:8080/v1/AUTH_test/test/foo/test.png` , i.e.

```
               Account: AUTH_test
             Container: test
                Object: foo/test.png
```

It should be quite trivial for the user to convert that to `swift://test/foo/test.png` before passing it to imgproxy. I'll document that.

Comment (joec4i) [reply transport/swift/swift.go]: That's implemented now with documentation update. (I just realised you suggested the same thing in the previous comment. Apologies for missing that.) Please have another look.

Comment (joec4i) [reply transport/swift/swift.go]: I rebased btw. Please check https://github.com/imgproxy/imgproxy/pull/837/commits/675f7bf1f99c9f5b2ac958c69568fe9d8b46c55e .

File: transport/swift/swift.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+		return &http.Response{
+			Status:     "400 Bad Request",
+			StatusCode: 400,
+			Proto:      "HTTP/1.1",
+			ProtoMajor: 1,
+			ProtoMinor: 1,
+			Close:      true,
+			Request:    req,
+		}, err
+	}
```
Comment (DarthSim) [inline transport/swift/swift.go:52]: This is useless now as we don't have an error to check

File: transport/swift/swift.go (hunk 5 above)
Comment (DarthSim) [reply transport/swift/swift.go:74]: Yup, I checked, we MUST close `object` here

Comment (joec4i) [reply transport/swift/swift.go:63]: yeah I added `defer object.Close()` at https://github.com/imgproxy/imgproxy/pull/837/files#diff-a187eae60042edf0861f826dce91206512effa21002f261eea57c5e19812a52fR55

File: transport/swift/swift.go (hunk 7 above)
Comment (joec4i) [reply transport/swift/swift.go]: 🤦 Nice catch. I forgot to clean that up. It's done now.

File: transport/swift/swift.go (hunk 5 above)
Comment (DarthSim) [reply transport/swift/swift.go:62]: With `defer object.Close()` in that place, we close `object` no matter what. So if we respond with 200, the response body will be closed. We should close `object` only when we respond with 304 as we don't pass it with the response.

**Comment by joec4i** [comment]
Hi @DarthSim Apologies for my carelessness. 😛  Please have another look. 🙏

File: transport/swift/swift.go (hunk 5 above)
Comment (joec4i) [reply transport/swift/swift.go:60]: You are right. I'm an idiot 😢 .  That's fixed now. Tested manually.

```
INFO    [2022-04-06T20:46:41+10:00] Completed in 545.22448ms /rs:fill/w:800/h:600/plain/swift://test/test.png  request_id=artKs1EMWHiGl3pncT0Fi method=GET status=200 client_ip="::1" image_url="swift://test/test.png" processing_options="Width: 800; Height: 600; Format: png"
INFO    [2022-04-06T20:46:42+10:00] Started /favicon.ico  request_id=eAj04jxtHh2BOtJqPjTFp method=GET client_ip="::1"
```

Comment (DarthSim) [reply transport/swift/swift.go:60]: Take it easy, could happen to everyone :)

**Comment by DarthSim** [comment]
Merged! Thanks for your contribution!

File: docs/serving_files_from_openstack_swift.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
+# Serving files from OpenStack Object Storage ("Swift")
+
+imgproxy can process images from OpenStack Object Storage, also known as Swift. To use this feature, do the following:
+
+1. Set `IMGPROXY_USE_SWIFT` environment variable to `true`
```
Comment (Travis-Turner) [inline docs/serving_files_from_openstack_swift.md:5]: Set IMGPROXY_USE_SWIFT... --> Set the IMGPROXY_USE_SWIFT...

File: docs/serving_files_from_openstack_swift.md [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+2. Configure Swift authentication with the following environment variables
```
Comment (Travis-Turner) [inline docs/serving_files_from_openstack_swift.md:6]: ...the following environment variables --> the following environment variables:

File: docs/serving_files_from_openstack_swift.md [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+   * `IMGPROXY_SWIFT_USERNAME`: the username for Swift API access. Default: blank
+   * `IMGPROXY_SWIFT_API_KEY`: the API key for Swift API access. Default: blank
+   * `IMGPROXY_SWIFT_AUTH_URL`: the Swift Auth URL. Default: blank
+   * `IMGPROXY_SWIFT_AUTH_VERSION`: the Swift auth version, set to 1, 2 or 3 or leave at 0 for autodetect.
+   * `IMGPROXY_SWIFT_TENANT`: the tenant name (optional, v2 auth only). Default: blank
+   * `IMGPROXY_SWIFT_DOMAIN`: the Swift domain name (optional, v3 auth only): Default: blank
+
+3. Use `swift://%{container}/%{object_path}` as the source image URL. e.g. an original object storage URL in the format of
```
Comment (Travis-Turner) [inline docs/serving_files_from_openstack_swift.md:14]: in the format of --> with the format of

File: docs/serving_files_from_openstack_swift.md [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+   `/v1/{account}/{container}/{object_path}` such as `http://127.0.0.1:8080/v1/AUTH_test/images/flowers/rose.jpg` should
```
Comment (Travis-Turner) [inline docs/serving_files_from_openstack_swift.md:15]: `/v1/{account}/{container}/{object_path}` such as `http://127.0.0.1:8080/v1/AUTH_test/images/flowers/rose.jpg` should --> `/v1/{account}/{container}/{object_path}`, such as `http://127.0.0.1:8080/v1/AUTH_test/images/flowers/rose.jpg`, should

**Comment by Travis-Turner** [comment]
> @Travis-Turner Could you check the docs, please?

Checked

**Comment by DarthSim** [comment]
Thanks a lot! Will fix it in master

**Pull request merged**